File: test/orchestrator.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { git } from "../src/git.js";
import { initialize } from "../src/init.js";
import { readJson, writeJson } from "../src/json.js";
import { planGitHubRelease } from "../src/orchestrator.js";
import type { ShipConfig, ShipManifest } from "../src/types.js";

const temporaryDirectories: string[] = [];

async function repository(): Promise<string> {
  const root = await fs.mkdtemp(path.join(os.tmpdir(), "smf-orchestrator-"));
  temporaryDirectories.push(root);
  await fs.mkdir(path.join(root, "ios"));
  await fs.writeFile(
    path.join(root, "pubspec.yaml"),
    "name: example\nversion: 1.0.0+1\n",
  );
  await fs.writeFile(path.join(root, "pubspec.lock"), "# fixture lockfile\n");
  await git(root, ["init", "-b", "main"]);
  await git(root, ["config", "user.name", "Test"]);
  await git(root, ["config", "user.email", "test@example.com"]);
  await git(root, ["add", "."]);
  await git(root, ["commit", "-m", "chore: bootstrap"]);
  await initialize({ root, bundleId: "dev.example.app" });
  await git(root, ["add", "."]);
  await git(root, ["commit", "-m", "chore: configure releases"]);
  return root;
}

afterEach(async () => {
  await Promise.all(
    temporaryDirectories
      .splice(0)
      .map((directory) => fs.rm(directory, { recursive: true, force: true })),
  );
});

describe("workflow routing", () => {
  it("routes only the configured pending release branch to candidate", async () => {
    const root = await repository();
    await git(root, ["checkout", "-b", "ship-my-flutter/ios"]);
    const manifestPath = path.join(root, ".ship-my-flutter", "manifest.json");
    const manifest = await readJson<ShipManifest>(manifestPath);
    manifest.platforms.ios = {
      ...manifest.platforms.ios,
      version: "1.1.0",
      pendingRelease: true,
    };
    await writeJson(manifestPath, manifest);
    await writeJson(path.join(root, ".ship-my-flutter", "changelog.json"), {
      schemaVersion: 1,
      platforms: {
        ios: {
          releases: {
            "1.1.0": {
              version: "1.1.0",
              preparedAt: "2026-07-26T00:00:00.000Z",
              baseSha: manifest.platforms.ios.baselineSha,
              headSha: manifest.platforms.ios.baselineSha,
              changes: [
                {
                  sha: manifest.platforms.ios.baselineSha,
                  type: "feat",
                  scope: "ios",
                  description: "Fixture release",
                  body: null,
                  breaking: false,
                  bump: "minor",
                  platforms: ["ios"],
                },
              ],
            },
          },
        },
      },
    });
    await git(root, ["add", "."]);
    await git(root, ["commit", "-m", "chore(ios): release 1.1.0"]);

    await expect(
      planGitHubRelease({
        root,
        github: { owner: "example", repo: "app", token: "unused" },
      }),
    ).resolves.toEqual({
      phase: "candidate",
      platform: "ios",
      version: "1.1.0",
      branch: "ship-my-flutter/ios",
    });

    await git(root, ["tag", "ios-v1.1.0"]);
    await expect(
      planGitHubRelease({
        root,
        github: { owner: "example", repo: "app", token: "unused" },
      }),
    ).resolves.toEqual({ phase: "noop" });
  });

  it("does nothing on unrelated branches", async () => {
    const root = await repository();
    await git(root, ["checkout", "-b", "docs"]);
    await expect(
      planGitHubRelease({
        root,
        github: { owner: "example", repo: "app", token: "unused" },
      }),
    ).resolves.toEqual({ phase: "noop" });
  });

  it("does nothing when iOS delivery is disabled", async () => {
    const root = await repository();
    const configPath = path.join(root, ".ship-my-flutter", "config.json");
    const config = await readJson<ShipConfig>(configPath);
    config.platforms.ios.enabled = false;
    await writeJson(configPath, config);
    await expect(
      planGitHubRelease({
        root,
        github: { owner: "example", repo: "app", token: "unused" },
      }),
    ).resolves.toEqual({ phase: "noop" });
  });
});

File: packages/smf_engine/test/orchestrator_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:smf_engine/smf_engine.dart';
import 'package:test/test.dart';

Future<Directory> repository() async {
  final root = await Directory.systemTemp.createTemp('smf-orchestrator-');
  addTearDown(() => root.delete(recursive: true));
  final origin = await Directory.systemTemp.createTemp(
    'smf-orchestrator-origin-',
  );
  addTearDown(() => origin.delete(recursive: true));
  await GitClient(root: origin.path).run(
    const <String>['init', '--bare', '-b', 'main'],
  );
  await Directory(p.join(root.path, 'ios')).create();
  await File(
    p.join(root.path, 'pubspec.yaml'),
  ).writeAsString('name: example\nversion: 1.0.0+1\n');
  await File(
    p.join(root.path, 'pubspec.lock'),
  ).writeAsString('# fixture lockfile\n');
  await GitClient(root: root.path).run(const <String>['init', '-b', 'main']);
  await GitClient(root: root.path).run(const <String>['config', 'user.name', 'Test']);
  await GitClient(root: root.path).run(const <String>[
    'config',
    'user.email',
    'test@example.com',
  ]);
  await GitClient(root: root.path).run(const <String>['add', '.']);
  await GitClient(root: root.path).run(const <String>['commit', '-m', 'chore: bootstrap']);
  await RepositoryInitializer.initialize(
    InitOptions(appRoot: root.path, iosBundleId: 'dev.example.app'),
  );
  await GitClient(root: root.path).run(const <String>['add', '.']);
  await GitClient(root: root.path).run(const <String>[
    'commit',
    '-m',
    'chore: configure releases',
  ]);
  await GitClient(root: root.path).run(<String>[
    'remote',
    'add',
    'origin',
    origin.path,
  ]);
  await GitClient(root: root.path).run(
    const <String>['push', '-u', 'origin', 'main'],
  );
  return root;
}

const context = GitHubContext(owner: 'example', repo: 'app', token: 'unused');

void main() {
  group('workflow routing', () {
    test(
      'routes only configured pending release branch to release-candidate',
      () async {
        final root = await repository();
        await GitClient(root: root.path).run(
          const <String>['checkout', '-b', 'smf/example/release'],
        );
        final paths = SmfPaths.resolve(root.path);
        final initial = await SmfState.manifest(root.path);
        final manifest = ManifestDto(
          schemaVersion: 1,
          platforms: ManifestPlatformsDto(
            ios: PlatformManifestDto(
              version: '1.1.0',
              endCommitHash: initial.platforms.ios.endCommitHash,
              isReleasePending: true,
            ),
            android: initial.platforms.android,
          ),
        );
        const encoder = JsonEncoder.withIndent('  ');
        await File(
          paths.manifest,
        ).writeAsString('${encoder.convert(manifest.toJson())}\n');
        final changelog = ChangelogDto(
          schemaVersion: 1,
          platforms: ChangelogPlatformsDto(
            ios: ChangelogPlatformDto(
              releases: <ChangelogPlatformReleaseVersionDto>[
                ChangelogPlatformReleaseVersionDto(
                  version: '1.1.0',
                  preparedAt: DateTime.utc(2026, 7, 26),
                  baseCommitHash: initial.platforms.ios.endCommitHash,
                  endCommitHash: initial.platforms.ios.endCommitHash,
                  changes: <ConventionalChangeDto>[
                    ConventionalChangeDto(
                      commitHash: initial.platforms.ios.endCommitHash,
                      type: 'feat',
                      scope: 'ios',
                      description: 'Fixture release',
                      body: null,
                      isBreaking: false,
                      versionBumpType: VersionBumpType.minor,
                      platforms: const <ReleasePlatform>[ReleasePlatform.ios],
                    ),
                  ],
                ),
              ],
            ),
            android: const ChangelogPlatformDto(
              releases: <ChangelogPlatformReleaseVersionDto>[],
            ),
          ),
        );
        await File(
          paths.changelog,
        ).writeAsString('${encoder.convert(changelog.toJson())}\n');
        await GitClient(root: root.path).run(const <String>['add', '.']);
        await GitClient(root: root.path).run(const <String>[
          'commit',
          '-m',
          'chore(ios): release 1.1.0',
        ]);

        final result = await const ReleaseOrchestrator().plan(
          workingDirectory: root.path,
          github: context,
        );
        expect(result.toJson(), <String, Object?>{
          'nextPhase': 'release-candidate',
          'targets': <Object?>[
            <String, Object?>{'platform': 'ios', 'version': '1.1.0'},
          ],
          'releaseBranch': 'smf/example/release',
        });
        expect(
          (await const ReleaseOrchestrator().plan(
            workingDirectory: root.path,
            github: context,
            selectedPlatform: ReleasePlatform.android,
          )).toJson(),
          <String, Object?>{'nextPhase': 'noop'},
        );

        await GitClient(root: root.path).run(
          const <String>['tag', 'example/ios-v1.1.0'],
        );
        await GitClient(root: root.path).run(
          const <String>['push', 'origin', 'example/ios-v1.1.0'],
        );
        expect(
          (await const ReleaseOrchestrator().plan(
            workingDirectory: root.path,
            github: context,
          )).toJson(),
          <String, Object?>{'nextPhase': 'noop'},
        );
      },
    );

    test('does nothing on unrelated branches', () async {
      final root = await repository();
      await GitClient(root: root.path).run(const <String>['checkout', '-b', 'docs']);
      expect(
        (await const ReleaseOrchestrator().plan(
          workingDirectory: root.path,
          github: context,
        )).toJson(),
        <String, Object?>{'nextPhase': 'noop'},
      );
    });

    test('does nothing when iOS delivery is disabled', () async {
      final root = await repository();
      await Directory(p.join(root.path, 'android')).create();
      final configFile = File(SmfPaths.resolve(root.path).config);
      await configFile.writeAsString('''
schema_version: 1
app_id: example
platforms:
  ios:
    enabled: false
  android:
    enabled: true
''');
      expect(
        (await const ReleaseOrchestrator().plan(
          workingDirectory: root.path,
          github: context,
        )).toJson(),
        <String, Object?>{'nextPhase': 'noop'},
      );
    });
  });
}
